ddprof-exporter/src/lib.rs

```diff
@@ -86,12 +91,29 @@ impl ProfileExporterV3 {
         tags: Vec<Tag>,
         endpoint: Endpoint,
     ) -> Result<ProfileExporterV3, Box<dyn Error>> {
-        Ok(Self {
+        let mut exporter = Self {
             exporter: Exporter::new()?,
             endpoint,
             family: family.into(),
             tags,
-        })
+        };
+        exporter.add_useful_tags();
+        Ok(exporter)
+    }
+
+    /// Add some useful tags
+    /// Currently only add Datadog-Container-ID if not already present
+    fn add_useful_tags(&mut self) {
+        let container_id_name = "Datadog-Container-ID";
+
+        if !self.tags.iter().any(|x| x.name == container_id_name) {
+            if let Some(container_id) = containerid::get_container_id() {
```

Comment:
Oh nice, pretty slick!  Let me see if I get this right.  If `get_container_id()` threw an error, then `Some(container_id)` will try to expand `Some(None)`, which will be false, and the tag will be skipped entirely.

The thing on my mind is, how will the user (or support engineers, like ourselves!) ever figure out the environmental factors that caused a container_id to be omitted?  What if we had a special "Datadog-Container-ID-errors" tag (or something) that gave a one-word description of why the "Datadog-Container-ID" couldn't be populated?  In the average case, we're sending along ~megabyte of data or more, and only once per minute, and almost always to the local TCP stack, so it's not like header minimization improves much, and it might have significant diagnostic value later.

No need to block on that, just an idea.

Reply:
If I understand Rust correctly `if let Some(container_id) = containerid::get_container_id()` is actually a kind of pattern matching. `get_container_id` returns an `Option<&str>`, ie. either `Some(&str)` or `None`.
`if` statement will be true only if `Some(&str)` is returned, this allows to both capture the returned value and do the pattern matching at the same time.

Reply:
Good remark on the error reporting. 
I am not sure that reporting errors through tags is a good idea, but perhaps it would be useful to have some king of local logging to be able to debug issues.
In `extract_container_id` I tried to distinguish between:
* IO errors: mainly /proc/self/cgroup not present, I don't see how we could fail to read it if it is present, or how it could contain invalid UTF8
* No container ID found (either because we don't run in a container or the layout in /proc/self/cgroup is unknown to us)
But this distinction is lost with `get_container_id`.

Reply:
Ah, indeed.  Putting an error in a tag does decouple the error from the source and it does deprive the calling application from doing _something_ about it, but interestingly this is a class of error which is not that useful to report to the profiler, but is very useful to be able to infer from the UX where the data is actually consumed.  From a diagnostic perspective, there's this weird interplay between the developer and the user.

Either way, I don't think this needs to be solved, and if it does need to be solved I don't think it needs to be done in this PR.  I agree with your classification of the different error modes, and I think the most general issue we'll run into is when the user is running in a system that has cgroup namespaces enabled (as does Ubuntu 21.04 IIRC).  In those cases, the host OS will probably tell us more than any particular error tag (if we know to look...)